File: 0123/dp.cpp
#include <algorithm>
#include <cassert>
#include <iostream>
#include <vector>

using namespace std;

constexpr int kTransactionCount = 2;

// https://leetcode.com/problems/best-time-to-buy-and-sell-stock-iii/discuss/135704/Detail-explanation-of-DP-solution
// dp[k, i] = max(dp[k, i-1], prices[i] - prices[j] + dp[k-1, j-1]), j=[0..i-1]
class Solution {
 public:
  int maxProfit(vector<int>& prices) {
    size_t size = prices.size();
    if (size == 0) return 0;
    vector<vector<int>> dp(kTransactionCount+1, vector<int>(size, 0));
    for (size_t k = 1; k <= kTransactionCount; k++) {
      int valley = prices[0];
      for (size_t i = 1; i < size; i++) {
        valley = min(valley, prices[i] - dp[k-1][i-1]);
        dp[k][i] = max(
          prices[i] - valley,   // sell
          dp[k][i-1]  // do nothing
        );
      }
    }
    return dp[kTransactionCount][size-1];
  }
};

int main() {
  Solution sol;
  vector<int> prices;

  prices = {3,3,5,0,0,3,1,4};
  assert(sol.maxProfit(prices) == 6);

  prices = {1,2,3,4,5};
  assert(sol.maxProfit(prices) == 4);

  prices = {7,6,4,3,1};
  assert(sol.maxProfit(prices) == 0);
}

File: 0123/dp.py
from typing import List


class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        if len(prices) == 0:
            return 0
        # dp[t, i] = max(dp[t, i-1],
        #                dp[t-1, j] - prices[j] + prices[i]) (0 <= j < n)
        num_transactions = 2
        dp = [[0 for _ in range(len(prices))]
              for _ in range(num_transactions+1)]
        for transaction in range(1, num_transactions+1):
            max_profit_when_buy = - 2 ** 32
            for day in range(1, len(prices)):
                max_profit_when_buy = max(
                    max_profit_when_buy,
                    dp[transaction-1][day-1] - prices[day-1]
                )
                dp[transaction][day] = max(
                    dp[transaction][day-1],
                    max_profit_when_buy + prices[day]
                )
        return dp[num_transactions][len(prices)-1]


if __name__ == '__main__':
    sol = Solution()

    prices = [3, 3, 5, 0, 0, 3, 1, 4]
    expected = 6
    assert sol.maxProfit(prices) == expected

    prices = [1, 2, 3, 4, 5]
    expected = 4
    assert sol.maxProfit(prices) == expected
